src/Microsoft.IdentityModel.Tokens/RsaCryptoServiceProviderProxy.cs

```diff
@@ -52,24 +56,20 @@ public RSACryptoServiceProviderProxy(RSACryptoServiceProvider rsa)
                 throw LogHelper.LogArgumentNullException("rsa");
 
             //
-            // If the provider does not understand SHA256, 
-            // replace it with one that does.
-            //
-            if (rsa.CspKeyContainerInfo.ProviderType != PROV_RSA_AES)
+            // Level up the provider type only if:
+            // 1. it is PROV_RSA_FULL or PROV_RSA_SCHANNEL which denote CSPs that only understand Sha1 algorithms
+            // 2. it is not associated with a hardware key
+            if ((rsa.CspKeyContainerInfo.ProviderType == PROV_RSA_FULL || rsa.CspKeyContainerInfo.ProviderType == PROV_RSA_SCHANNEL) && !rsa.CspKeyContainerInfo.HardwareDevice)
```

Comment:
because we don't support Sha1 in Wilson. For sha256, sha384, sha512 (the algorithms that we support in wilson), we need to uplevel. And in desktop WCF, we don't support Sha384 and Sha512, so uplevel only for sha256